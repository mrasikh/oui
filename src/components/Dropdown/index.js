import React from 'react';
import classNames from 'classnames';
import PropTypes from 'prop-types';
import { Manager, Popper, Target } from 'react-popper';
import { withState } from 'recompose';
import { withToggle } from '../../utils/recompose-utils';

import DropdownContents from './DropdownContents';
import DropdownListItem from './DropdownListItem';
import DropdownBlockLink from './DropdownBlockLink';
import DropdownBlockLinkText from './DropdownBlockLinkText';
import DropdownBlockLinkSecondaryText from './DropdownBlockLinkSecondaryText';

@withToggle
@withState('overChildren', 'setOverChildren', false)
class Dropdown extends React.Component {
  static displayName = 'Dropdown';

  handleOnBlur = () => {
    if (!this.props.overChildren) {
      this.props.hide();
    }
  };

  handleToggle = () => {
    this.props.setOverChildren(false);
    this.props.toggle();
  };

  onMouseOver = () => this.props.setOverChildren(true);

  onMouseLeave = () => this.props.setOverChildren(false);

  render() {
    const {
      activator,
      arrowIcon,
      buttonContent,
      children,
      isDisabled = false,
      isOpen,
      displayError = false,
      fullWidth,
      placement = 'bottom-start',
      style,
      testSection,
      width = 200,
      zIndex = 999,
    } = this.props;

    const groupClass = classNames(
      'oui-dropdown-group',
      {['width--1-1']: fullWidth},
      {['oui-form-bad-news']: displayError},
      {['is-active']: isOpen},
    );

    const buttonClass = classNames(
      'oui-button',
      {[`oui-button--${style}`]: style},
      {['oui-button--full soft--left text--left']: fullWidth}
    );

    const iconClass = classNames(
      'push-half--left',
      {
        ['oui-arrow-inline--down']: arrowIcon === true || arrowIcon === 'down',
        ['oui-arrow-inline--left']: arrowIcon === 'left',
        ['oui-arrow-inline--right']: arrowIcon === 'right',
        ['oui-arrow-inline--up']: arrowIcon === 'up',
      }
    );

    return (
      <Manager
        data-ui-component={ true }
        className={ groupClass }
        data-test-section={ testSection }>
        <Target>
          {
            buttonContent && (
              <button
                type='button'
                className={ buttonClass }
                disabled={ isDisabled }
                onClick={ this.handleToggle }
                onBlur={ this.handleOnBlur }>
                <div className='flex'>
                  <div className='flex--1 truncate'>{ buttonContent }</div>
                  {
                    !!arrowIcon && arrowIcon !== 'none' && (
                      <div className='text--right'><span className={ iconClass }/></div>
                    )
                  }
                </div>
              </button>
            )
          }
          {
            activator && React.cloneElement(activator, {
              // trigger the dropdown if the child element is clicked on
              onClick: this.handleToggle,
              onBlur: this.handleOnBlur,
            })
          }
        </Target>
        <Popper
          placement={ placement }
          className='oui-dropdown-children'
          style={{
            zIndex: zIndex,
            position: 'absolute',
            width: width,
            marginTop: 2,
            marginBottom: 2,
            borderRadius: 'var(--border-radius)',
            boxShadow: '0 2px 3px rgba(0,0,0,.1)',
          }}
          onMouseOver={ this.onMouseOver }
          onMouseLeave={ this.onMouseLeave }
          onClick={ this.handleToggle }>
          { isOpen && !isDisabled && children }
        </Popper>
      </Manager>
    );
  }
}

Dropdown.propTypes = {
  /** React element that when clicked activates the dropdown
   * Either this prop OR buttonContent should be used, not both
   */
  activator: PropTypes.node,
  /** Arrow icon direction:
    * - Defaults to 'none', which hides the arrow
    * - passing a prop value of false also hides the arrow
    * - passing a prop value of true uses down arrow
    */
  arrowIcon: PropTypes.oneOf([
    'down',
    'left',
    'none',
    'right',
    'up',
  ]),
  /** Button text, can be a string or element.
   * Either this prop OR activator should be used, not both
   */
  buttonContent: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.element,
  ]),
  /** Dropdown contents, typically using the <Blocklist> component. */
  children: PropTypes.node.isRequired,
  /** Show error state. */
  displayError: PropTypes.bool,
  /** Button width is either full or inline-block. */
  fullWidth: PropTypes.bool,
  /** Unused... */
  handleClick: PropTypes.func,
  hide: PropTypes.func,
  /** Disable button. */
  isDisabled: PropTypes.bool,
  isOpen: PropTypes.bool,
  /* Whether or not the children are currently moused over */
  overChildren: PropTypes.bool,
  /** Popper placement property */
  placement: PropTypes.oneOf([
    'top',
    'top-start',
    'top-end',
    'bottom',
    'bottom-start',
    'bottom-end',
    'right',
    'right-start',
    'right-end',
    'left',
    'left-start',
    'left-end',
  ]),
  /* Control for overChildren value */
  setOverChildren: PropTypes.bool,
  /** Button style, e.g. highlight, danger, outline. */
  style: PropTypes.string,
  /** For automated testing only. */
  testSection: PropTypes.string,
  /** Toggle control for isOpen. */
  toggle: PropTypes.func,
  /** Dropdown menu width, in pixels. */
  width: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]),
  /** Override default dropdown menu z-index. */
  zIndex: PropTypes.number,
};

Dropdown.defaultProps = {
  arrowIcon: 'none',
};

Dropdown.Contents = DropdownContents;
Dropdown.ListItem = DropdownListItem;
Dropdown.BlockLink = DropdownBlockLink;
Dropdown.BlockLinkText = DropdownBlockLinkText;
Dropdown.BlockLinkSecondaryText = DropdownBlockLinkSecondaryText;

export default Dropdown;
