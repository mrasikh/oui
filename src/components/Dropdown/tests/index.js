import React from 'react';
import Dropdown from '../index';
import { mount, shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';

const data = [
  {title: 'Manual', description: 'Dolcelatte cheeseburger swiss paneer cow gouda edam cheese slices'},
  {title: 'Maximize Conventions', header: 'Automation',
    description: 'Everyone loves cut the cheese cow squirty cheese emmental jarlsberg halloumi camembert de normandie.'},
  {title: 'Faster Results', description: 'Cauliflower cheese stilton paneer hard cheese fondue.'},
];

describe('components/Dropdown', () => {
  it('should render children when isOpen is true', () => {
    const component = mount(
      <Dropdown
        icon={ true }
        isOpen={ true }
        buttonContent='Dropdown'>
        <ul>
          { data.map((item, index) => {
            return (
              <li key={ index }>{ item.title }</li>
            );
          })
          }
        </ul>
      </Dropdown>
    );

    expect(component.find('Dropdown').props().isOpen).toBe(true);
    expect(component.find('ul').exists()).toBe(true);
  });

  it('should not render children when isDisabled is true', () => {
    const component = shallow(
      <Dropdown
        icon={ true }
        buttonContent='Dropdown'
        idDisabled={ true }>
        <ul>
          { data.map((item, index) => {
            return (
              <li key={ index }>{ item.title }</li>
            );
          })
          }
        </ul>
      </Dropdown>
    );
    expect(shallowToJson(component)).toMatchSnapshot();
  });

  it('should call the child onClick handler when clicked', () => {
    const onClickMock = jest.fn();
    const component = shallow(
      <Dropdown
        icon={ true }
        buttonContent='Dropdown'
        idDisabled={ true }>
        <ul>
          { data.map((item, index) => {
            return (
              <li
                key={ index }
                onClick={ onClickMock }
                data-test-section={ `dropdown-item-${index}` }>
                { item.title }
              </li>
            );
          })
          }
        </ul>
      </Dropdown>
    );

    component.find('[data-test-section="dropdown-item-1"]').simulate('click');
    expect(onClickMock.mock.calls.length).toBe(1);
  });

  it('should not use .oui-arrow-inline--down when icon isEqual to triangle', () => {
    const component = shallow(
      <Dropdown
        icon={ true }
        buttonContent='Dropdown'>
        <ul>
          { data.map((item, index) => {
            return (
              <li key={ index }>{ item.title }</li>
            );
          })
          }
        </ul>
      </Dropdown>
    );
    expect(shallowToJson(component)).toMatchSnapshot();
  });

  it('should use oui-button--full class when fullWidth is true', () => {
    const component = shallow(
      <Dropdown
        icon={ true }
        buttonContent='Dropdown'
        fullWidth={ true }>
        <ul>
          { data.map((item, index) => {
            return (
              <li key={ index }>{ item.title }</li>
            );
          })
          }
        </ul>
      </Dropdown>
    );
    expect(shallowToJson(component)).toMatchSnapshot();
  });

  it('should highlight button class when style equals highlight', () => {
    const component = shallow(
      <Dropdown
        icon={ true }
        buttonContent='Dropdown'
        style='highlight'>
        <ul>
          { data.map((item, index) => {
            return (
              <li key={ index }>{ item.title }</li>
            );
          })
          }
        </ul>
      </Dropdown>
    );
    expect(shallowToJson(component)).toMatchSnapshot();
  });

  it('should include oui-form-bad-news class when display error is true', () => {
    const component = shallow(
      <Dropdown
        buttonContent='Dropdown'
        displayError={ true }>
      </Dropdown>
    );
    expect(shallowToJson(component)).toMatchSnapshot();
  });

  it('should render an activator node when passed as a prop', () => {
    const component = shallow(
      <Dropdown
        activator={ <button>Click me</button> }
        displayError={ true }>
      </Dropdown>
    );
    expect(shallowToJson(component)).toMatchSnapshot();
  });
});
