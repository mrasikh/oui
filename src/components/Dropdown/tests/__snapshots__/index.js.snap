// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Dropdown should highlight button class when style equals highlight 1`] = `
<withHandlers(withState(Dropdown))
  arrowIcon="none"
  buttonContent="Dropdown"
  icon={true}
  isOpen={false}
  style="highlight"
  toggle={[Function]}
>
  <ul>
    <li
      key="0"
    >
      Manual
    </li>
    <li
      key="1"
    >
      Maximize Conventions
    </li>
    <li
      key="2"
    >
      Faster Results
    </li>
  </ul>
</withHandlers(withState(Dropdown))>
`;

exports[`components/Dropdown should include oui-form-bad-news class when display error is true 1`] = `
<withHandlers(withState(Dropdown))
  arrowIcon="none"
  buttonContent="Dropdown"
  displayError={true}
  isOpen={false}
  toggle={[Function]}
/>
`;

exports[`components/Dropdown should not render children when isDisabled is true 1`] = `
<withHandlers(withState(Dropdown))
  arrowIcon="none"
  buttonContent="Dropdown"
  icon={true}
  idDisabled={true}
  isOpen={false}
  toggle={[Function]}
>
  <ul>
    <li
      key="0"
    >
      Manual
    </li>
    <li
      key="1"
    >
      Maximize Conventions
    </li>
    <li
      key="2"
    >
      Faster Results
    </li>
  </ul>
</withHandlers(withState(Dropdown))>
`;

exports[`components/Dropdown should not use .oui-arrow-inline--down when icon isEqual to triangle 1`] = `
<withHandlers(withState(Dropdown))
  arrowIcon="none"
  buttonContent="Dropdown"
  icon={true}
  isOpen={false}
  toggle={[Function]}
>
  <ul>
    <li
      key="0"
    >
      Manual
    </li>
    <li
      key="1"
    >
      Maximize Conventions
    </li>
    <li
      key="2"
    >
      Faster Results
    </li>
  </ul>
</withHandlers(withState(Dropdown))>
`;

exports[`components/Dropdown should render an activator node when passed as a prop 1`] = `
<withHandlers(withState(Dropdown))
  activator={
    <button>
      Click me
    </button>
  }
  arrowIcon="none"
  displayError={true}
  isOpen={false}
  toggle={[Function]}
/>
`;

exports[`components/Dropdown should use oui-button--full class when fullWidth is true 1`] = `
<withHandlers(withState(Dropdown))
  arrowIcon="none"
  buttonContent="Dropdown"
  fullWidth={true}
  icon={true}
  isOpen={false}
  toggle={[Function]}
>
  <ul>
    <li
      key="0"
    >
      Manual
    </li>
    <li
      key="1"
    >
      Maximize Conventions
    </li>
    <li
      key="2"
    >
      Faster Results
    </li>
  </ul>
</withHandlers(withState(Dropdown))>
`;
